File: java/QuickSortJava/src/main/java/com/github/insanusmokrassar/KotlinVSJava/JavaComparisonModule/QuickSortJava/QuickSorter.java
package com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.QuickSortJava;

public class QuickSorter {
    public static void sort(Integer[] source) {// to is exclude
        sort(source, 0, source.length);
    }


    public static void sort(Integer[] source, int from, int to) {// to is exclude
        if (to - from < 2) {
            return;
        }

        int rightIndex = to - 1;

        int mainIndex = QuickSort.random.nextInt(to - from - 1) + from;

        int main = source[mainIndex];

        int leftIndex = from;

        while (leftIndex < mainIndex) {
            int current = source[leftIndex];
            if (current >= main) {
                if (rightIndex == mainIndex) {
                    source[rightIndex] = current;
                    source[leftIndex] = source[mainIndex - 1];

                    mainIndex--;
                    source[mainIndex] = main;
                } else {
                    int right = source[rightIndex];
                    source[rightIndex] = current;
                    source[leftIndex] = right;
                }
                rightIndex--;
            } else {
                leftIndex++;
            }
        }

        while (mainIndex < rightIndex) {
            int current = source[rightIndex];
            if (current < main) {
                source[mainIndex] = current;
                source[rightIndex] = source[mainIndex + 1];

                mainIndex++;
                source[mainIndex] = main;
            } else {
                rightIndex--;
            }
        }

        sort(source, from, mainIndex);
        sort(source, mainIndex + 1, to);
    }
}

File: kotlin/QuickSortKotlin/src/main/kotlin/com/github/insanusmokrassar/KotlinVSJava/KotlinComparisonModule/QuickSortKotlin/QuickSorter.kt
package com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.QuickSortKotlin

fun sort(source: Array<Int>, from: Int = 0, to: Int = source.size) {
    if (to - from < 2) {
        return
    }

    var rightIndex = to - 1

    var mainIndex = random.nextInt(to - from - 1) + from

    val main = source[mainIndex]

    var i = from

    while (i < mainIndex) {
        val current = source[i]
        if (current >= main) {
            if (rightIndex == mainIndex) {
                source[rightIndex] = current
                source[i] = source[mainIndex - 1]
                source[mainIndex - 1] = main

                mainIndex--
                rightIndex--
            } else {
                val right = source[rightIndex]
                source[rightIndex] = current
                source[i] = right

                rightIndex--
            }
        } else {
            i++
        }
    }

    sort(source, from, mainIndex)
    sort(source, mainIndex + 1, to)
}
